drivers: mfd: add support for MAX2221X multi-function device

csarip opened the pull request:
Add initial support for the MAX22216/MAX22217 device. These devices have four 36 V half-bridges that support various drive modes, supported by MFD, PWM, and MISC drivers.

csarip committed:
drivers: mfd: add support for MAX2221X multi-function device

Implements SPI-based register access (read, write, update) for the
MAX22216/MAX22217 multi-function device.

Signed-off-by: Cherrence Sarip <cherrence.sarip@analog.com>

csarip committed:
drivers: pwm: add support for MAX2221X pwm

Implements PWM driver support for the MAX22216/MAX22217 device.
This driver integrates with the Zephyr PWM API and uses the parent
MFD device to perform register access.

Signed-off-by: Cherrence Sarip <cherrence.sarip@analog.com>

csarip committed:
drivers: misc: add support for MAX2221X misc

Add MISC driver support for the MAX22216/MAX22217 device.
The driver provides miscellaneous functions and uses the parent
MFD device for register access.

Signed-off-by: John Erasmus Mari Geronimo <johnerasmusmari.geronimo@analog.com>
Signed-off-by: Cherrence Sarip <cherrence.sarip@analog.com>

MaureenHelm commented on drivers/pwm/Kconfig.max2221x:
```diff
@@ -0,0 +1,20 @@
+	help
+	  Analog Devices MAX22216/MAX22217 PWM driver initialization priority.
+
+#endif #PWM_MAX2221X
```
Is there a reason not to use `PWM_INIT_PRIORITY`?

If you do need a driver-specific init priority, you need to uncomment the if/endif statements and convert `config PWM_MAX2221X` to a menuconfig

MaureenHelm commented on drivers/pwm/pwm_max2221x.c:
```diff
@@ -0,0 +1,329 @@
+		return -EINVAL;
+	}
+
+	*duty_cycle = (uint16_t)((pulse * 0x7FFFU) / period);
```
```suggestion
	*duty_cycle = (uint16_t)((pulse * UINT16_MAX) / period);
```

MaureenHelm commented on drivers/pwm/pwm_max2221x.c:
```diff
@@ -0,0 +1,329 @@
+	*freq_divisor = divisors[0];
+	min_difference = abs((int)(user_freq_hz - (master_freq / divisors[0])));
+
+	for (int i = 1; i < 4; i++) {
```
```suggestion
	for (int i = 1; i < ARRAY_SIZE(divisors); i++) {
```

MaureenHelm reviewed (changes requested):
Thanks @csarip!

csarip commented:
v2:

- added suggested changes in pwm_max2221x.c and misc Kconfig
- converted config PWM_MAX2221X to a menuconfig
- added an entry in tests/drivers/build_all/pwm/testcase.yaml

MaureenHelm commented on tests/drivers/build_all/pwm/testcase.yaml:
```diff
@@ -59,3 +59,5 @@ tests:
   drivers.pwm.rts5912.build:
     platform_allow: rts5912_evb
+  drivers.pwm.max2221x.build:
+    platform_allow: nucleo_l476rg
```
Need to add
```yaml
    extra_args: DTC_OVERLAY_FILE=max2221x.overlay
```

MaureenHelm commented:
Needs rebase

csarip commented:
v5:

- Added minor changes from the checks.

@MaureenHelm @kartben

csarip commented:
v5:

- Added minor changes from the doxygen checks.
- Rebased onto the latest from the zephyrproject-rtos/zephyr main branch.

@MaureenHelm @kartben

csarip commented:
v6:

- Added minor changes from the doxygen checks.
- Rebased onto the latest zephyrproject-rtos/zephyr main branch.

Could someone help trigger the pending checks? I’d really appreciate it. Thank you! 🙂

@MaureenHelm @kartben
